Support uploading user profile photos

spelak-salesforce opened the pull request:
# Critical Changes

# Changes

# Issues Closed

davisagli commented on cumulusci/tasks/salesforce/users.py:
```diff
@@ -0,0 +1,135 @@
+        path = self.options["photo_path"]
+
+        self.logger.info(f"Setting user photo to {path}")
+        with open(path, "rb") as version_data:
```
I'd recommend turning the string path into a pathlib.Path instance: `path = pathlib.Path(self.options["photo_path"])`.

Check `path.exists()` to confirm the file exists and show a helpful error if not.

And then pathlib.Path has some useful attributes and methods to use below:
- `path.name` instead of `os.path.split(path)[-1]`
- `path.stem` instead of `os.path.splitext(os.path.split(path)[-1])[0]`
- `path.read_bytes()` to get the file contents in a single call without needing to use `open` as a context manager.

spelak-salesforce replied:
Will do.

spelak-salesforce replied:
pushed changes to use`pathlib`.   This is much simpler!  ✅

spelak-salesforce commented on cumulusci/cumulusci.yml:
```diff
@@ -464,13 +464,21 @@ tasks:
+    upload_user_profile_photo:
+        group: Data Operations
+        description: Uploads a profile photo for the User whose user_field equals user_field_value.
+        class_path: cumulusci.tasks.salesforce.UploadUserProfilePhoto
```
Yes.   I've been weighing that question, and I like the key/value option.  I could even take a list of key/value pairs which translate to a list of `AND` filters for the User SOQL query.

davisagli replied:
Hmm, maybe just better to take the entire SOQL WHERE clause as an string option instead of trying to invent a one-off, incomplete way of building SOQL expressions?

spelak-salesforce replied:
I have an idea that I think works well.   I check the field's [SOAP Type](https://developer.salesforce.com/docs/atlas.en-us.apexcode.meta/apexcode/apex_enum_Schema_SOAPType.htm).   If the SOAP Type "text" meaning STRING (`xsd:string`), ID (`tns:ID`), or ADDRESS (`urn:address`), I wrap the field value in single quotes.  Else, don't wrap the field in single quotes. 

It seems pretty effective so the query supports Booleans and even date/datetime literals.

davisagli replied:
Accepting the whole WHERE clause as a string would also support those things, as well as supporting everything else that's valid in a WHERE clause (OR, IN, LIKE, comparison operators, checks for equality to null...).

I'm not asking you to implement support for those things. I'm sure you could. But I don't think generic logic for building a SOQL expression from a Python structure belongs in this task, and I don't see a particular reason we need it.

Any thoughts, @davidmreed?

spelak-salesforce replied:
I agree it's more generic, and I will update to support a `filter` SOQL rather than `filters`.    My heart liked how the implementations would look.   

```yml
tasks:
    dev_upload_profile_photo_grace:
        group: "SFDO-Grants: Internal config and storytelling"
        description: Uploads a profile photo for (business account) grantseeker Grace Walker.
        class_path: cumulusci.tasks.salesforce.users.UploadProfilePhoto
        options:
            photo: dev/photos/grace.png
            filters:
                FirstName: Grace
                LastName: Walker
```

spelak-salesforce replied:
+1 `where` option give the most flexibility plus it seems simpler to implement

davidmreed replied:
I agree with @davisagli. The WHERE clause is a well-established CCI pattern that fits here. Supporting a kind of SOQL builder in the options has merit, but I think we should evaluate that option as part of a bigger rethink of how options work across tasks rather than introduce a one-off mechanic that will need to be maintained as backwards-compatible.

spelak-salesforce replied:
implemented and pushed.

spelak-salesforce commented on cumulusci/tasks/salesforce/__init__.py:
```diff
@@ -42,6 +42,8 @@
     "UpdateProfile": "cumulusci.tasks.salesforce.update_profile",
     "UpdateAdminProfile": "cumulusci.tasks.salesforce.update_profile",
+    "UploadUserProfilePhoto": "cumulusci.tasks.salesforce.users",
+    "UploadDefaultUserProfilePhoto": "cumulusci.tasks.salesforce.users",
```
Will update.

spelak-salesforce replied:
I created a sub-directory of `salesforce` called `users` so tasks are from `cumulusci.tasks.salesforce.users`.   My hunch is we will eventually add more user administration tasks later.

davisagli commented on cumulusci/tasks/salesforce/users/tests/test_photos.py:
```diff
@@ -0,0 +1,46 @@
+        self.task.logger = mock.Mock()
+        self.task.logger.info._expected_calls = [
+            mock.call(f"Setting user photo to {path}")
+        ]
```
I would be inclined to write this as an integration test that uses a real file on disk and asserts that the expected ContentVersion data was prepared, rather than a strict unit test. I try to reserve use of mocks for cases where the tests need to interact with an external service or would otherwise be slow.

spelak-salesforce replied:
Will do.   I haven't touched the test since the re-factor.

spelak-salesforce replied:
The tests are integration tests in regards to files.  `pathlib` is not mocked.  Only simple salesforce is mocked.

spelak-salesforce commented on cumulusci/tasks/salesforce/users/photos.py:
```diff
@@ -0,0 +1,178 @@
+
+        user_ids = []
+        try:
+            for record in self.sf.query_all(query)["records"]:
```
I did that for my first pass.  But then I considered logging the User IDs to help debug any problems of why the query isn't selective enough.

What are your thoughts on the following log message: 

```
"More than one User found (at least 2): {}".format(", ".join(user_ids))
```

davisagli replied:
Sure, that sounds handy.

spelak-salesforce commented on cumulusci/tasks/salesforce/users/photos.py:
```diff
@@ -0,0 +1,161 @@
+
+        user_ids = []
+        try:
+            for record in self.sf.query_all(query)["records"]:
```
We have a `LIMIT 2`.  Does it matter using `self.sf.query` vs `self.sf.query_all`?

davidmreed replied:
I don't believe it does; there won't be a nextRecordsUrl ever.

spelak-salesforce commented:
@davisagli  @davidmreed 

This is ready for review though I still need to update the Pull Request body:
- Task written ✅  
- Added task to `cumulusci/cumulusci.yml` ✅ 
- Task has `task_docs` + ran `make docs` + reviewed the docs look as expected ✅ 
- Fully tested.   Only mocked simple salesforce.  Did not mock `pathlib` and manually created files, etc. ✅ 
- Updated Pull Request body 📆 
- Want to perform manual tests after completing unit testing.  📆

spelak-salesforce commented:
> @davisagli @davidmreed
> 
> This is ready for review though I still need to update the Pull Request body:
> 
> * Task written ✅
> * Added task to `cumulusci/cumulusci.yml` ✅
> * Task has `task_docs` + ran `make docs` + reviewed the docs look as expected ✅
> * ~Fully tested~.   Only mocked simple salesforce.  Did not mock `pathlib` and manually created files, etc. ✅
> * Updated Pull Request body 📆
> * Want to perform manual tests after completing unit testing.  📆
> 
> ## Update
> pytest passes when ran in the test directory but fails when ran in the root directory. I need to update getting reference to my test `photo_mock.txt` file. Will update Monday.

@davisagli @davidmreed 

This PR is now fully ready for review.

spelak-salesforce commented on cumulusci/tasks/salesforce/users/tests/test_photos.py:
```diff
@@ -0,0 +1,372 @@
+                temp_photo_path.is_file()
+            ), "Path the the photo mock in the temporary directory should point to a file"
+
+            self.task.sf.ContentVersion.create._expected_calls = [
```
The intention is to attach the expected called right to the mock to keep everything in one variable.  I've found it useful when there are several mocked methods.  But since this creates confusion, I won't reuse this technique and probably create separate local variables.
